### PR TITLE
Add new built-in flag `--dump-help`

### LINKED ISSUE #164: Provide a way to dump the help text for all commands and options in JSON
This can be helpful to write tooling that generates documentation files in various formats.

### PR DESCRIPTION
**Pull request by KS1019**
This PR resolves #164 and adds a new built-in flag `--dump-help` to dump all the help information in JSON format. This feature can be useful when parsing help information and generating documentation for a command. I will add documentation if necessary. In that case, I would appreciate it if anyone could direct me to the right place to add the documentation.

### Checklist
- [x] I've added at least one test that validates that my change is working, if appropriate
- [x] I've followed the code style of the rest of the project
- [x] I've read the [Contribution Guidelines](CONTRIBUTING.md)
- [ ] I've updated the documentation if necessary

### CODE REVIEW

**Comment by KS1019** [comment]
@natecook1000 Please take a look when you have time. Thank you!

**Review by natecook1000 — COMMENTED** [review]
Looks like a good start, @KS1019! I haven't gotten a chance to review `DumpHelpInfoGenerator` yet, but wanted to give you a couple notes here.

File: Sources/ArgumentParser/Usage/DumpHelpInfoGenerator.swift [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+// Copyright (c) 2020 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+//
+//===----------------------------------------------------------------------===//
+@_implementationOnly import Foundation
+
+internal struct DumpHelpInfoGenerator {
+  struct CommandInfo: Codable {
```
Comment (natecook1000) [inline Sources/ArgumentParser/Usage/DumpHelpInfoGenerator.swift:14]: I think `CommandInfo` and `ArgumentInfo` will probably be independently useful — can you move them out to be top-level types?

File: Sources/ArgumentParser/Usage/DumpHelpInfoGenerator.swift [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+      // Used only for arguments and options
+      var isRequired: Bool?
+      var defaultValue: String?
+      var valueName: String?
+      
+      // Used only for subcommands
+      var isDefault: Bool?
+    }
+    
+    var command: ArgumentInfo
```
Comment (natecook1000) [inline Sources/ArgumentParser/Usage/DumpHelpInfoGenerator.swift:49]: Let's not re-use this for the command's metadata — should be okay to just put the command-specific details right here instead.

File: Sources/ArgumentParser/Usage/HelpGenerator.swift [hunk 3]
```diff
@@ -185,6 +185,7 @@ internal struct HelpGenerator {
         optionElements.append(element)
       }
     }
+    optionElements.append(.init(label: "--dump-help", abstract: "Dump help information."))
```
[the remark this answers is not in the document]
Comment (KS1019) [reply Sources/ArgumentParser/Usage/HelpGenerator.swift]: Fixed at 40de584

File: Sources/ArgumentParser/Usage/DumpHelpInfoGenerator.swift (hunk 1 above)
Comment (KS1019) [reply Sources/ArgumentParser/Usage/DumpHelpInfoGenerator.swift]: Addressed at 94a81ed

File: Sources/ArgumentParser/Usage/DumpHelpInfoGenerator.swift (hunk 2 above)
Comment (KS1019) [reply Sources/ArgumentParser/Usage/DumpHelpInfoGenerator.swift]: Fixed at 9c8e5a7

File: Sources/ArgumentParser/Parsing/CommandParser.swift [hunk 4]
```diff
@@ -78,6 +80,11 @@ extension CommandParser {
     guard !split.contains(anyOf: self.commandStack.getHelpNames()) else {
       throw HelpRequested()
     }
+    
+    // Look for `--dump-help`
+    guard !split.contains(Name.long("dump-help")) else {
+      throw DumpHelpInfoRequested()
+    }
```
[the remark this answers is not in the document]
Comment (KS1019) [reply Sources/ArgumentParser/Parsing/CommandParser.swift]: At 4d97d0c, I modified the code to throw CommandError like `--version` does. I will look into more if I can remove DumpHelpInfoCommand completely.

**Comment by natecook1000** [comment]
@swift-ci Please test

**Comment by KS1019** [comment]
Hi @natecook1000
I fixed codes that had compilation errors and made changes to follow the upstream changes. Please take a look again when you have time!

**Comment by natecook1000** [comment]
@swift-ci Please test

**Comment by natecook1000** [comment]
@swift-ci Please test Linux platform

**Comment by KS1019** [comment]
Hi @natecook1000

It seems that tests are failing on the Linux Platform because specifying only `encoder.outputFormatting = .prettyPrinted` gives me same JSON but in different order depending on platforms.

I am considering adding `.sortedKeys` to `encoder.outputFormatting` but since they are available only on macOS 10.13 or newer, I am not sure if that's the best option.

Alternatively, I think we can change how we test the generated JSON from comparing them as strings to comparing them as values by converting them back to the codable values.

I would love to know what do your think the best way would be to solve the problem!

**Comment by natecook1000** [comment]
@KS1019 I think it would be fine to round trip through JSON and check for equality with the expectation that way. String equality for JSON is definitely tricky due to its order independence. Thanks!

**Comment by KS1019** [comment]
@natecook1000 
I've made changes to address the issues with testing JSONs. Please take a look when you have time!

**Comment by natecook1000** [comment]
@swift-ci Please test

**Comment by compnerd** [comment]
This regressed the Windows build :-( (#336 should repair that)

**Pull request merged**